### PR TITLE
Cortexm asm to c use cmsis

### PR DESCRIPTION
**Pull request by manochavikas**
This patchset uses the cmsis functions to replace some of the assembly coding done in reset.S for cortex-m architectures. It brings more code clarity and the benefit of easy code maintenance as cmsis repo takes care of intrinsic functions. 

We should be able to move all the stuff from reset.S to PrepC, let's keep it for next patch.

Cheers,
Vikas

### CODE REVIEW

File: arch/arm/core/cortex_m/prep_c.c [hunk 1]
```diff
@@ -24,6 +24,45 @@
...
+	__disable_irq();
+#elif defined(CONFIG_ARMV7_M_ARMV8_M_MAINLINE)
+	__set_BASEPRI(_EXC_IRQ_DEFAULT_PRIO);
+#else
+#error Unknown ARM architecture
+#endif /* CONFIG_ARMV6_M_ARMV8_M_BASELINE */
+}
+
+#ifdef CONFIG_INIT_STACKS
+static inline void init_stacks(void)
```
Comment (ioannisg) [inline arch/arm/core/cortex_m/prep_c.c]: I would rather define `init_stacks()` only if `CONFIG_INIT_STACKS `is defined; no need of empty function definition for a function that is only used in this file.
This means, of course that init_stacks shall only be invoked `#ifdef CONFIG_INIT_STACKS`.

File: arch/arm/core/cortex_m/prep_c.c [hunk 2]
```diff
@@ -24,6 +24,45 @@
 #include <arch/arm/cortex_m/cmsis.h>
 #include <string.h>
 
+extern uint32_t _interrupt_stack;
```
Comment (ioannisg) [inline arch/arm/core/cortex_m/prep_c.c]: u32_t

Comment (ioannisg) [inline arch/arm/core/cortex_m/prep_c.c]: Also, I wonder whether we could include stack.h here (so we get the declaration of _interrupt_stack)

File: arch/arm/core/cortex_m/prep_c.c [hunk 3]
```diff
@@ -24,6 +24,45 @@
...
+
+static inline void switch_sp_to_psp(void)
+{
+	__set_CONTROL(__get_CONTROL() | CONTROL_SPSEL_Msk);
+	/* Synchronous barrier to make psp new SP effective immediately */
+	__ISB();
+}
+
+static inline void set_and_switch_to_psp(void)
```
Comment (ioannisg) [inline arch/arm/core/cortex_m/prep_c.c]: perhaps `set_and_switch_to_use_psp`?

File: arch/arm/core/cortex_m/prep_c.c [hunk 4]
```diff
@@ -24,6 +24,45 @@
 #include <arch/arm/cortex_m/cmsis.h>
 #include <string.h>
 
+extern uint32_t _interrupt_stack;
+
+static inline void switch_sp_to_psp(void)
+{
+	__set_CONTROL(__get_CONTROL() | CONTROL_SPSEL_Msk);
+	/* Synchronous barrier to make psp new SP effective immediately */
```
Comment (ioannisg) [inline arch/arm/core/cortex_m/prep_c.c]: I would copy the previous explanatory text here

**Review by ioannisg — CHANGES_REQUESTED** [review]
Thanks for the patch :)
Looks good - some  comments from my side

File: arch/arm/core/cortex_m/prep_c.c (hunk 1 above)
Comment (manochavikas) [reply arch/arm/core/cortex_m/prep_c.c]: Sure, I will update it in next pull request.

File: arch/arm/core/cortex_m/prep_c.c (hunk 2 above)
Comment (manochavikas) [reply arch/arm/core/cortex_m/prep_c.c]: Agree.

Comment (manochavikas) [reply arch/arm/core/cortex_m/prep_c.c]: I will check it.

File: arch/arm/core/cortex_m/prep_c.c (hunk 3 above)
Comment (manochavikas) [reply arch/arm/core/cortex_m/prep_c.c]: both switch & use would be redundant...

File: arch/arm/core/cortex_m/prep_c.c (hunk 4 above)
Comment (manochavikas) [reply arch/arm/core/cortex_m/prep_c.c]: ok.

File: arch/arm/core/cortex_m/prep_c.c (hunk 3 above)
Comment (ioannisg) [reply arch/arm/core/cortex_m/prep_c.c]: Well, I was thinking of "switching from using MSP to using PSP"
So it made sense to me :)

File: arch/arm/core/cortex_m/prep_c.c [hunk 5]
```diff
@@ -23,6 +23,44 @@
...
+	 * When changing the stack pointer, software must use an ISB instruction
+	 * immediately after the MSR instruction. This ensures that instructions
+	 * after the ISB instruction execute using the new stack pointer.
+	 */
+	__ISB();
+}
+
+static inline void set_and_switch_to_psp(void)
+{
+	uint32_t process_sp = (uint32_t)&_interrupt_stack;
```
Comment (ioannisg) [inline arch/arm/core/cortex_m/prep_c.c]: u32_t (and in all similar places)
Also, you might be getting a warning for missing an empty line between declaration of variable (in line 41) and line 42

File: arch/arm/core/cortex_m/prep_c.c (hunk 1 above)
Comment (ioannisg) [reply arch/arm/core/cortex_m/prep_c.c]: Looks like yo added the #ifdef, no?

File: arch/arm/core/cortex_m/prep_c.c (hunk 5 above)
Comment (manochavikas) [reply arch/arm/core/cortex_m/prep_c.c]: my bad, let me fix all uint32_t to u32_t in v3.
No, no such warning.

File: arch/arm/core/cortex_m/prep_c.c (hunk 3 above)
Comment (manochavikas) [reply arch/arm/core/cortex_m/prep_c.c]: Great ! Thanks for your review comments, appreciate it.

**Comment by manochavikas** [comment]
Hi,

On Tue, Jan 29, 2019 at 1:02 PM Ioannis Glaropoulos <
notifications@github.com> wrote:

> *@ioannisg* commented on this pull request.
> ------------------------------
>
> In arch/arm/core/cortex_m/prep_c.c
> <https://github.com/zephyrproject-rtos/zephyr/pull/12802#discussion_r252015950>
> :
>
> > +	switch_sp_to_psp();
> +}
> +
> +void lock_interrupts(void)
> +{
> +#if defined(CONFIG_ARMV6_M_ARMV8_M_BASELINE)
> +	__disable_irq();
> +#elif defined(CONFIG_ARMV7_M_ARMV8_M_MAINLINE)
> +	__set_BASEPRI(_EXC_IRQ_DEFAULT_PRIO);
> +#else
> +#error Unknown ARM architecture
> +#endif /* CONFIG_ARMV6_M_ARMV8_M_BASELINE */
> +}
> +
> +#ifdef CONFIG_INIT_STACKS
> +static inline void init_stacks(void)
>
> Looks like yo added the #ifdef, no?
>

Yes, as per your review comment.

Cheers,
Vikas

File: arch/arm/core/cortex_m/prep_c.c (hunk 5 above)
Comment (manochavikas) [reply arch/arm/core/cortex_m/prep_c.c]: you were right, just ran checkpatch which gave me the warning.
I will fix it in v3.

**Review by ioannisg — APPROVED** [review]

**Comment by carlescufi** [comment]
recheck

**Review by agross-oss — APPROVED** [review]

**Comment by ioannisg** [comment]
@manochavikas you might want to look at #12878, as well; it is a similar challenge. Eventually we want to use CMSIS instead of assembly as much as possible.

**Pull request merged**